src/components/dialog-full-screen/dialog-full-screen.test.tsx

```diff
@@ -0,0 +1,432 @@
+import React, { useState } from "react";
+import { render, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import StyledFullScreenHeading from "../../__internal__/full-screen-heading/full-screen-heading.style";
+import DialogFullScreen from "./dialog-full-screen.component";
+import StyledContent from "./content.style";
+import StyledIconButton from "../icon-button/icon-button.style";
+import { StyledHeader, StyledHeading } from "../heading/heading.style";
+import Form from "../form";
+import { StyledFormContent } from "../form/form.style";
+import CarbonProvider from "../carbon-provider";
+import Logger from "../../__internal__/utils/logger";
+
+// mock Logger.deprecate so that no console warnings occur while running the tests
+const loggerSpy = jest.spyOn(Logger, "deprecate");
+
+const ControlledDialog = ({
+  onCancel,
+  startOpen = true,
+}: {
+  onCancel: () => void;
+  startOpen?: boolean;
+}) => {
+  const [open, setOpen] = useState(startOpen);
+  return (
+    <DialogFullScreen
+      open={open}
+      onCancel={() => {
+        setOpen(false);
+        onCancel();
+      }}
+    />
+  );
+};
+
+beforeAll(() => {
+  loggerSpy.mockImplementation(() => {});
+  jest.useFakeTimers();
+});
+
+afterAll(() => {
+  loggerSpy.mockRestore();
+  jest.useRealTimers();
+});
+
+test("dialog element has aria-modal attribute set to true when open", () => {
+  render(
+    <CarbonProvider>
+      <DialogFullScreen open />
+    </CarbonProvider>
+  );
+
+  expect(screen.getByRole("dialog")).toHaveAttribute("aria-modal", "true");
+});
+
+test("open dialog element does not have aria-modal attribute set to true when role is not `dialog`", () => {
+  render(
+    <CarbonProvider>
+      <DialogFullScreen open role="main" />
+    </CarbonProvider>
+  );
+
+  expect(screen.getByRole("main")).not.toHaveAttribute("aria-modal");
+});
+
+test("the contentRef prop is forwarded to the dialog content", () => {
+  const mockRef = {
+    current: null,
+  };
+
+  render(<DialogFullScreen open contentRef={mockRef} />);
+
+  expect(mockRef?.current).toBe(
+    screen.getByTestId("dialog-full-screen-content")
+  );
+});
+
+test("the dialog container should be focused when the dialog opens", () => {
+  render(
+    <DialogFullScreen open>
+      <input type="text" />
+    </DialogFullScreen>
+  );
+
+  jest.runAllTimers();
+
+  expect(screen.getByRole("dialog")).toHaveFocus();
+});
```

Comment:
**suggestion**: for the focus-related tests in this file, what do you think about replacing the use of `jest.runAllTimers` with `waitFor`? This would be more closely aligned to RTL's recommendations on handling async behaviour:

```suggestion
test("the dialog container should be focused when the dialog opens", async () => {
  render(
    <DialogFullScreen open>
      <input type="text" />
    </DialogFullScreen>
  );

  await waitFor(() => {
    expect(screen.getByRole("dialog")).toHaveFocus();
  });
});
```

Reply:
happy to do this for this test, but note that for the next one where we're testing the `disableAutoFocus` prop, we shouldn't use `waitFor` as that won't fail if the prop's functionality ever gets broken - the dialog is always initially unfocused (it only becomes so after a delay) so "waiting" for it not to be focused will always pass. That one only fails in the broken case by using `runAllTimers`. I'll add a clarifying comment to that test to explain it.

Reply:
I see your point about ensuring the `disableAutoFocus` functionality is tested. However, I believe using `waitFor` would be more appropriate and idiomatic with RTL in this case.

The goal is to verify that the dialog remains unfocused and does not receive focus after any potential UI updates or asynchronous behaviour completes, rather than simply checking if the `disableAutoFocus` code path is executed.

That said, I’m not strongly opposed to the current approach if you feel it fits this specific scenario here. My only concern is that, in general, we avoid tying our tests too closely to implementation details.

Reply:
> The goal is to verify that the dialog remains unfocused and does not receive focus after any potential UI updates or asynchronous behaviour completes, rather than simply checking if the disableAutoFocus code path is executed.

Exactly. This isn't about coverage at all, and I'm not sure where I mis-spoke to give you that impression. It's about ensuring the test works.

Using `waitFor(() => expect(screen.getByRole("dialog")).not.toBeFocused()` doesn't do what you rightly say the goal is, because it is saying "keep checking for a short period and if the dialog is focused for that whole time, fail the test". But, by design, the dialog is not focused at the point the `waitFor` starts, so this won't ever fail.

What we want is to say "keep checking for a short period and if the dialog ever moves from being unfocused to being focused, then fail the test". Unfortunately I don't know of a good way to do that - `runAllTimers` isn't great because it's very implementation-specific (because we happen to use `setTimeout` I think deep inside the code that does the focusing), so I'd love to have a better way, but `waitFor` isn't designed for these situations. That's for "I know this isn't true right now, but it should be in a little bit", which is ideal for tests like this specific one you've commented on where we want to check the dialog gets focused. It doesn't help with "this is actually true right now, but I want to make sure it becomes false in a little bit", which is what we're testing in the `disableAutoFocus` case.

Hope this makes more sense now, but please let me know if not!

Reply:
Ah thanks for clarifying this @robinzigmond! I now see what you mean about the assertion being satisfied immediately due to the dialog initially having focus. Happy to stick with this 👍🏼